### ISSUE TITLE
Enhancement: Support for Lua-based bundle definitions (porter.lua)

### ISSUE THREAD

**Issue by jdolitsky** [getporter/porter#173]
# TL;DR
Porter should accept Lua-based bundle definitions in addition to (or instead of) YAML.

# Summary
The primary goal of Porter (as I understand) is to enhance the CNAB authoring experience. 

For this reason, I want to suggest that the Porter client is extended to support Lua-based bundle definitions. Since the format is more "complex", the default behavior for `porter build` should check for the presence of a file named `porter.lua` and evaluate it, otherwise fall back on `porter.yaml` (default).

# Explanation
Lua-based definitions require less effort to modify and require less code duplication compared to YAML.

## Example of Lua being used in a similar setting
I would honestly be skeptical of supporting something like this myself if I hadn't seen it come in handy in some real-life scenarios.

Checkout the [GoFish](https://gofi.sh/index.html) project, which defines "fish food" package definitions in Lua. Because of this, modifications are ultra-streamlined. Here is an example:

(snip from https://github.com/fishworks/fish-food/blob/master/Food/delve.lua)
```lua
local name = "delve"
local version = "1.1.0"

-- Executable name is actually "dlv" (vs. "delve")
local exe = "dlv"

-- One .zip contains executables for all 3 platforms,
-- so the url and sha256 fields are commom across all packages
local url = "https://dl.bintray.com/jetbrains/golang/com/jetbrains/delve/" .. version .. "/delve-" .. version .. ".zip"
local sha256 = "dbeeda31b44fa6e7f65078a350427bac2414fc301cb9ad2b99e8f695b597580c"

food = {
    name = name,
    description = "Debugger for the Go programming language",
    license = "MIT",
    homepage = "https://github.com/derekparker/delve",
    version = version,
    packages = {
        {
            os = "darwin",
            arch = "amd64",
            url = url,
            sha256 = sha256,
            resources = {
                {
                    path = exe .. "/mac/" .. exe,
                    installpath = "bin/" .. exe,
                    executable = true
                }
            }
        },
...
```

When a new version of delve is released, potentially all that we will need to change is just the `version` and the `sha256` variables. In addition, other parts of the definition are generated/interpolated using variables (such as `url`). These are things that are not easily done in YAML.

## Example of `porter.lua`
Porter can take a similar approach to the fish food example by requiring and evaluating a top-level/global `bundle` variable in `porter.lua`.

Take the following `porter.yaml`:
```yaml
name: cloud-communicator
version: 0.1.0
description: "this thing talks to the cloud. no joke."

invocationImage: r.mysite.io/trogdor-burninator/cloud-communicator:0.1.0

mixins:
  - exec

install:
  - description: "Install my-bundle"
    exec:
      command: /usr/local/bin/talk-to-the-cloud
      arguments:
        - --setup
        - --provider
        - aws
        - -f
        - aws-setup.conf

uninstall:
  - description: "Uninstall my-bundle"
    exec:
      command: /usr/local/bin/talk-to-the-cloud
      arguments:
        - --teardown
        - --provider
        - aws
        - -f
        - aws-teardown.conf
```

The equivalent `porter.lua` might look like the following:
```lua
local name = "cloud-communicator"
local version = "0.1.0"

local registryHost = "r.mysite.io"
local registryRepo = "trogdor-burninator/" .. name

local command = "/usr/local/bin/talk-to-the-cloud"
local provider = "aws"
local installArgs = {"--setup", "--provider", provider, "-f", provider .. "-setup.conf"}
local uninstallArgs = {"--teardown", "--provider", provider, "-f", provider .. "-teardown.conf"}

bundle = {
    name =  name,
    version = version,
    description = "this thing talks to the cloud. no joke.",
    invocationImage = registryHost .. "/" .. registryRepo .. ":" .. version,
    mixins = {"exec"},
    install = {
        {
            description = "Install " .. name,
            exec = {command = command, arguments = installArgs}
        }
    },
    uninstall = {
        {
            description = "Uninstall " .. name,
            exec = {command = command, arguments = uninstallArgs}
        }
    }
}
```

This is a very basic example. Given the concept of custom Porter mixins, there is an endless stream of possibilities for how Lua might be used to build a bundle with Porter.

## YAML, Lua, or both?

For now, I suggest supporting both. Given the basic nature of the bundle definition, although YAML definitions may appear more verbose than the Lua counterpart, both options allow the user to provide all the required fields (metadata, invocation image, mixins, etc.).

I do think, however, that the Porter documentation should shift focus to Lua in any examples etc. as to promote its usage and to share the benefits that it provides for bundle authors.

I imagine that if the Lua option becomes the primary option for bundle authors, and new feature requests cannot be achieved in YAML, it might be discussed as a long-term option to deprecate YAML completely in the next major release of Porter.

## Other considerations
The biggest cons to this are that this introduces a large dependency into the project. So long as that dependency does its job well, I don't think that introducing this would actually require much additional end-user support (besides some good initial documentation).

Any issues that arise would likely be related to how to use the Lua language itself. Both the YAML document and the evaluated Lua `bundle` object would/should be unmarshalled into the same Go type.

In order to allow for something like this, Porter would need to support the Lua language using a pure Go implementation.

Helm (also a Go-based project) is expected to ship v3 with Lua support. There is active development being done on the underlying library that Helm is planning to use, [Azure/golua](https://github.com/Azure/golua).

GoFish currently uses [yuin/gopher-lua](https://github.com/yuin/gopher-lua) which appears to handle the fish food use-case quite well.

I'd suggest an implementation of this use Azure/golua, primarily due to the fact that Helm will eventually rely on it. This will provide Porter with a long-term support model for adding such a dependency.

---

Thanks for reading. Look forward to hearing your thoughts. Please take your time :)
cc @carolynvs @jeremyrickard

**Comment by carolynvs** [comment]
We have been having a bit of trouble with yaml parsing lately, so maybe @jeremyrickard won't need too much encouragement to support another language. 

Actually it's always been on the backlog to support multiple markup languages, #37, it's just a bit farther out on our roadmap. Hardock and Fisher are both on our team, so we are pretty aware of gofish (which we already want to make use of for mixin distribution) and the new lua engine. Overall, this is a pretty interesting idea!

I'm not sure how many people are comfortable with lua, but the same could be said for hcl or toml which were also on the list of languages we had on the list of support. 😀

**Comment by technosophos** [comment]
The fact that @jdolitsky referenced Trogdor the Burninator pretty much means we MUST do this. ;-)

In all honestly, I am intrigued by the idea. The nice thing about YAML is that it can be easily audited (since it is declarative). But Lua would make building complex Porters much easier.

**Comment by jdolitsky** [comment]
Glad to see there is some good initial reaction :)

@carolynvs did not see #37 (was searching issues for "lua"). Nice! Lua can be one of the many options.

I'd be happy to take this forward with the green light. Could add a common parser interface to support any future languages, such as:
```
Parser interface {
    Unmarshall(content []byte) (*Bundle, error)
}
```

Then implement it with a `YAMLParser` and `LuaParser` (`TOMLParser`, `HCLParser`, etc. in the future).

**Comment by garethr** [comment]
I generally like this, but if you know my predilection for small languages you already knew that. Here's a few thoughts that are likely worth considering as part of going from 1 to n. These aren't really specific to porter, but relevant to the general domain I think.

* I think it's always worth having a lowest common denominator data format. It makes integration with arbitrary tooling and workflows easier. No-one wants to see folks data templating Lua or other DSL code.
* Should these be implemented in tree or as stand alone plugins? This is mainly worth considering with a view to the question "if someone sent a good PR to add support for XParser, would we accept it?"

Other random notes:

* [Tengo](https://github.com/d5/tengo) appears a very nice fit for embedded Go languages at the moment
* [StarLark](https://github.com/google/starlark-go) (formerly Skylark) is well suited to this usecase too
* Their is now a port of the [Puppet language parser in Go](https://github.com/lyraproj/puppet-parser) too

One negative thing that comes from multiple input formats is the potential for making it hard to test generically. It would be _very_ useful to have a command and public method on the interface to get the raw data, maybe output to JSON? That's a good hook for generic testing tools or other workflow orientated tooling, without having to be concerned with the variety of potential inputs.

**Comment by jdolitsky** [comment]
@garethr 

> * I think it's always worth having a lowest common denominator data format

Agree. JSON probably.

> * Should these be implemented in tree or as stand alone plugins?

I think plugins might be the actual right thing. Can we somehow re-use mixins installation process to provide Porter "language packs"? Lua, HCL, and others ship as language packs. Community can contribute new packs. Do things need to be written in Go at this point? How did you picture this?

**Comment by carolynvs** [comment]
Just to clarify my earlier comment, support for other languages in the porter manifest is very far down the line on the porter roadmap. I don't want to stop people from dreaming big!  Just setting expectations on where we are focusing our efforts right now.

That said I O U a roadmap and more milestones, so I'll try to do that today. 😊

**Comment by jdolitsky** [comment]
I just got the Lua->YAML conversion working with Azure/golua here:
https://github.com/jdolitsky/goluamapper/tree/master/examples/lua_to_yaml

**Comment by carolynvs** [comment]
@jdolitsky Are you thinking of converting from lua to yaml as part of the parsing process? I thought that it would go directly from lua to the go structs? I think I missed a step in your design. 😁

**Comment by jdolitsky** [comment]
@carolynvs you're right, sorry if that was misleading. The goluamapper package is a fork of yuin/gluamapper used by GoFish to convert a Lua table to a Go struct. The "lua_to_yaml" example is just a proof we can go from one to the other. 

Here is an example of using goluamapper to go from Lua table to Go struct:
https://github.com/jdolitsky/goluamapper/blob/master/examples/readme/main.go

So it really depends how this will look. If adding "first-class" Lua support to Porter (my initial suggestion), it would look something like the latter example. In a "language pack plugin" scenario, it may be the case that external plugins need to hand off YAML/JSON to Porter to deal with.

**Comment by jdolitsky** [comment]
I've just put together a simple tool called "lupo" that converts a `porter.lua` into `porter.yaml` and then runs Porter: https://github.com/jdolitsky/lupo

We can use this as a way to prove out the concept for now.

**Comment by carolynvs** [comment]
Closing, supporting other languages than YAML is no longer on the roadmap.
